stdlib/internals.ncl

```diff
@@ -38,6 +38,11 @@
           %seal% sealing_key (%chng_pol% label) value,
 
   "$forall" = fun sealing_key polarity contract label value =>
+    let flip = match {
+      `Positive => `Negative,
+      `Negative => `Positive,
+    } in
+    let polarity = if %dualize% label then flip polarity else polarity in
```

Comment:
So you flip the polarity if the label is dualized, although the polarity was already flipped as well in the code generating the recursive environment. This means the polarity is back to the original one? Is there any difference between a normal label and a dualized one, then? I might be a bit lost.

Reply:
`polarity` here is not the current polarity in the label. It's the statically assigned polarity of the `forall` binding during contract generation. Setting `dualize` to true in the label basically makes the code that `subcontract` generated pretend that it was called with an initial polarity of `Negative`.

Reply:
Ah, if I understand correctly, you flip two different polarities when dualizing: the current polarity of the label (makes sense, as the whole contract moves to the codomain), and the one corresponding to what was the polarity of the `forall` when it was introduced. I'm not sure to understand why we need the second flip. Isn't `T.dualize()` in the notes defined as the codomain contract of `T -> Dyn` ? Shouldn't the foralls that are already introduced stay the same?

Reply:
When `subcontract` generates the domain contract for `T -> Dyn`, it calls `T.subcontract(vars.clone(), pol.flip(), sy)`. So the polarities of all the `forall`'s in `T` are flipped. That behaviour is replicated with the check here.

Reply:
But you already flip the label polarity when dualizing, right? There is a call to `flip()`. This should corresponds to the `pol.flip()`. So, any `forall` introduced inside `T` should inherit the flipped polarity and shouldn't need an additional flip, it seems.

Additionally, here, the dualization seems to also targets type variables that are already introduced in `T` (free in `T`, the ones corresponding to `vars` inside `subcontract`, which aren't affected by the codomain contract flip). The fact that it works is still a bit confusing, but there might be a "double dualization error that cancels out" as well. Or I might be misunderstanding something, but I'm having a hard time mapping the formalization of `T.dualize()` to this dualization operation here.

Reply:
Let's walk through an example, say `T = forall a. a`. This is translated, by `T.subcontract(pol = `Positive)`, into the contract
```
$forall 0 `Positive ($forall_var 0)
```
The contract `T.dualize()` should then be the domain contract in `T -> Dyn`, so we need to look at what `T -> Dyn` is translated into. That is
```
$func (T.subcontract(pol = `Negative)) $dyn
== $func ($forall 0 `Negative ($forall_var 0)) $dyn
```
So `T.dualize()` must be the contract `$forall 0 `Negative ($forall_var 0)`. Note that no label polarities have entered the picture yet, but still the polarity (statically) passed to `$forall` is flipped.

Now, when the resulting contracts are applied, another layer of polarity tracking is initiated in the label. Let's imagine we're applying the contract `T -> Dyn` to a function `f` taking a `T`. Then the following reductions take place, where I'm pretending `%go_codom%` and `%go_dom%` primops don't do anything:
```
   ($func ($forall 0 `Negative ($forall_var 0)) $dyn) label f
== fun x => $dyn label (f ($forall 0 `Negative ($forall_var 0) (chng_pol label) x))
== fun x => f ($forall 0 `Negative ($forall_var 0) (chgn_pol label) x)
```

When building the recursive environment in a record, when thinking of a recursive field as a function `T -> Dyn`, a recursive reference `x` should then be patched to
```
$forall 0 `Negative ($forall_var 0) (chgn_pol label) x'
```
where `x'` is `x` with the original `T` contract applied to it.

Reply:
Yes, I think I get it when the `forall` is introduced inside `T`. What I'm having a hard time seeing is how this works when the `forall` is introduced outside of the record. So, something like the record map type (say a specialized version): `forall a. (a -> a) -> {_: a} -> {_: a}`. Here, the contract `T` is mostly `a`. Shouldn't the corresponding `forall` contract then stay the same, instead of having its polarity flipped?

Reply:
In some sense, it feels like it treats `forall a. a` and `a` exactly the same, which is bugging me

Reply:
The corresponding `forall` contract will stay unflipped, but the reason is a bit convoluted. Let's say we're dealing with a type like `forall a. {_: a} -> {_: a}` and let's see which contracts actually make it into the recursive environment. At first, the contract generation code will produce
```
$forall 0 `Positive ($func ($dyn_record ($forall_var 0)) ($dyn_record ($forall_var 0)))
```
Evaluation will then proceed and `$dyn_record` will insert just `$forall_var 0` into the pending contracts of the record that it is applied to. So when building a recursive environment, there are no `$forall`'s left in the contract term whose polarity would be flipped. Instead, the original polarity of the `forall` is recorded in the typing environment of the label and stays untouched.

In a sense, setting the `dualize` flag in the label to true flips only the `forall`'s that have not yet been introduced. That's also what is checked by the new test in `tests/integration/pass/contracts.ncl`.

Reply:
Ooooh, it's the `$forall` contract, not the `$forall_var`, in this snippet.... Here goes my wrong axiom :sweat_smile: . Now this absolutely makes sense. Thanks for taking the time to explain.

It's a really non trivial magic trick we're pulling here, we'll to document it very clearly and in details, if we want this to be remotely understandable 6 months from now. The examples you posted here are valuable and should be stored somewhere, before we put the note in good shape.